Modules/BarcodeScanner/BarcodeScannerView.swift

```diff
@@ -0,0 +1,320 @@
+
+            guard self.captureSession.canAddOutput(self.captureMetadataOutput) else {
+                return DispatchQueue.main.async { [weak self] in
+                    self?.model.onError(.videoCapturingConfigurationFailed)
+                }
+            }
+
+            self.captureSession.addOutput(self.captureMetadataOutput)
+            self.captureMetadataOutput.metadataObjectTypes = self.model.barcodeTypes.map { $0.rawValue }
+            self.captureMetadataOutput.setMetadataObjectsDelegate(self, queue: .main)
+            self.isCaptureSessionConfigured = true
+
+            DispatchQueue.main.async { [weak self] in
+                self?.setVideoOrientation(to: UIApplication.shared.statusBarOrientation)
+                self?.didChangeState()
+            }
+        }
+    }
+
+    public override func didChangeModel() {
+        super.didChangeModel()
+
+        captureSessionQueue.async { [weak self] in
+            guard let self = self, self.isCaptureSessionConfigured else { return }
+
+            self.captureSession.beginConfiguration()
+            defer { self.captureSession.commitConfiguration() }
+
+            self.captureMetadataOutput.metadataObjectTypes = self.model.barcodeTypes.map { $0.rawValue }
+        }
+    }
+
+    public override func willMove(toSuperview newSuperview: UIView?) {
+        switch (superview, newSuperview) {
+        case (.none, .some):
+            isScanningEnabled = true
+
+        case (.some, .none):
+            isScanningEnabled = false
+
+        default:
+            break
+        }
+    }
+
+    public override func layoutSubviews() {
+        super.layoutSubviews()
+
+        captureVideoPreviewLayer.frame = bounds
+    }
+
+    private func didChangeState() {
+        if isScanning, isScanningEnabled, !isCaptureSessionRunning {
+            self.captureSessionQueue.async { [weak self] in
+                guard let self = self else { return }
+
+                self.addObservers()
+                self.captureSession.startRunning()
+                self.isCaptureSessionRunning = self.captureSession.isRunning
+            }
+        }
+
+        if (!isScanning || !isScanningEnabled), isCaptureSessionRunning {
+            self.captureSessionQueue.async { [weak self] in
+                guard let self = self else { return }
+
+                self.captureSession.stopRunning()
+                self.isCaptureSessionRunning = self.captureSession.isRunning
+                self.removeObservers()
+            }
+        }
+    }
+
+    /// Starts the scan process for the given `barcodeTypes` in the `model`.
+    public func startScanning() {
+        guard !isScanning else { return }
+
+        isScanning = true
+
+        requestVideoCapturingAuthorization { [weak self] accessGranted in
+            if accessGranted {
+                self?.captureSessionQueue.resume()
+            } else {
+                self?.model.onError(.videoCapturingNotAuthorized)
```

Comment:
Should we show an error here or just retry / try to request the authorisation by calling `requestVideoCapturingAuthorization`?

Reply:
If the video capturing is not authorized doing a request would land in the same location so an error is better here.